### PR TITLE
SoftMax OP 

### PR DESCRIPTION
**Pull request by lambda7xx**
**Description of changes:**
- SoftMax OP 

**Related Issues:**

Linked Issues:
- Issue #1079 

Issues closed by this PR:
- Closes #

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1106)

### CODE REVIEW

File: lib/kernels/include/kernels/softmax_kernels.h [hunk 1]
```diff
@@ -1,26 +1,27 @@
...
-public:
-  SoftmaxPerDeviceState(FFHandler handle,
-                        Softmax const *softmax,
-                        Legion::Domain const &input_domain);
+// Note(lambda): SoftmaxPerDeviceState may need add more elements
+struct SoftmaxPerDeviceState {
+  PerDeviceFFHandle handle;
   ffTensorDescriptor_t inputTensor;
-  bool profiling;
-  int dim;
```
Comment (reyna-abhyankar) [inline lib/kernels/include/kernels/softmax_kernels.h:15]: I think `dim` is still needed.

File: lib/runtime/src/ops/softmax.cc [hunk 2]
```diff
@@ -50,367 +54,172 @@ SoftmaxParams Softmax::get_params() const {
...
+
+  ParallelTensorShape output_shape = get_output_shape(attrs, input.shape);
+
+  SimTaskBinding init_binding;
+  // Note: what should init_binding?
+  init_binding.bind(INPUT, input);
+  init_binding.bind(OUTPUT, output_shape);
+  init_binding.bind_arg(ATTRS, attrs);
+  init_binding.bind_arg(PROFILING, settings);
+  init_binding.bind_arg(HANDLE, ff_handle());
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/softmax.cc:168]: I think `init_binding` only needs this

File: lib/runtime/src/ops/softmax.cc [hunk 3]
```diff
@@ -50,367 +54,172 @@ SoftmaxParams Softmax::get_params() const {
...
-    Layer const *layer,
-    std::vector<ParallelTensor> const &inputs) {
-  long long value;
-  layer->get_int_property("softmax_dim", value);
-  int dim = (int)value;
-  return new Softmax(model,
-                     inputs[0],
-                     (inputs[0]->num_dims - 1 - dim) % inputs[0]->num_dims,
-                     layer->name);
+  binding.bind_arg(HANDLE, ff_handle());
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/softmax.cc:60]:
```suggestion
  binding.bind_arg(HANDLE, ff_handle());
  binding.bind_arg(ATTRS, attrs);
```

**Review by reyna-abhyankar — CHANGES_REQUESTED** [review]

File: lib/runtime/src/ops/softmax.cc [hunk 4]
```diff
@@ -35,382 +39,177 @@ using Legion::TaskLauncher;
...
-#define DIMFUNC(DIM)                                                           \
-  case DIM:                                                                    \
-    return forward_task_with_dim<DIM>(task, regions, ctx, runtime);
-    LEGION_FOREACH_N(DIMFUNC)
-#undef DIMFUNC
-    default:
-      assert(false);
-  }
+static optional<float> forward_task_impl(TaskArgumentAccessor const &acc) {
+  auto input = acc.get_tensor<Permissions::RO>(A_INPUT);
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/softmax.cc:93]:
```suggestion
  auto input = acc.get_tensor<Permissions::RO>(INPUT);
```

File: lib/runtime/src/ops/softmax.cc [hunk 5]
```diff
@@ -35,382 +39,177 @@ using Legion::TaskLauncher;
...
-  launcher.add_field(1, FID_DATA);
-  runtime->execute_index_space(ctx, launcher);
+static optional<float> backward_task_impl(TaskArgumentAccessor const &acc) {
+  ProfilingSettings profiling = acc.get_argument<ProfilingSettings>(PROFILING);
+
+  auto input_grad = acc.get_tensor_grad<Permissions::RW>(INPUT);
+  auto input = acc.get_tensor<Permissions::RO>(INPUT);
+  assert(input_grad.shape == input.shape);
+
+  auto output_grad = acc.get_tensor<Permissions::RO>(OUTPUT);
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/softmax.cc:122]:
```suggestion
  auto output_grad = acc.get_tensor_grad<Permissions::RO>(OUTPUT);
```

File: lib/runtime/src/ops/softmax.cc [hunk 6]
```diff
@@ -35,382 +39,177 @@ using Legion::TaskLauncher;
...
-        cost_metrics.forward_time);
-  }
-  // Free softmaxmeta
-  delete m;
-  return true;
+template <>
+void register_task<SOFTMAX_FWD_TASK_ID>() {
+  OpTaskSignature fwd(OpTaskType::FWD);
+
+  fwd.add_arg_slot<bool>(PROFILING);
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/softmax.cc:198]:
```suggestion
  fwd.add_arg_slot<ProfilingSettings>(PROFILING);
```

**Review by reyna-abhyankar — DISMISSED** [review]

**Review by lambda7xx — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/flexflow/1106)* status: 0 of 5 files reviewed, 6 unresolved discussions (waiting on @lockshaw and @reyna-abhyankar)

___
*[`lib/kernels/include/kernels/softmax_kernels.h` line 0 at r1](https://reviewable.io/reviews/flexflow/flexflow/1106#gh-1343085227:-NgzRMaF4PSKWGRweEeA:b-896fix) ([raw file](https://github.com/flexflow/flexflow/blob/675a22b33920974b0b5feaae2ae78f3364088a54/lib/kernels/include/kernels/softmax_kernels.h#L0)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/flexflow/FlexFlow/pull/1106#discussion_r1343085227) to the right spot — sorry!)*

I think `dim` is still needed. 
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/softmax.cc` line 60 at r1](https://reviewable.io/reviews/flexflow/flexflow/1106#gh-1343088111:-NgzRO_W5eD7xlwJGdue:b-896fix) ([raw file](https://github.com/flexflow/flexflow/blob/675a22b33920974b0b5feaae2ae78f3364088a54/lib/runtime/src/ops/softmax.cc#L60)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

```suggestion
  binding.bind_arg(HANDLE, ff_handle());
  binding.bind_arg(ATTRS, attrs);
```
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/softmax.cc` line 168 at r1](https://reviewable.io/reviews/flexflow/flexflow/1106#gh-1343087453:-NgzRjgrC_-FXXpR9qmB:b-896fix) ([raw file](https://github.com/flexflow/flexflow/blob/675a22b33920974b0b5feaae2ae78f3364088a54/lib/runtime/src/ops/softmax.cc#L168)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

I think `init_binding` only needs this
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/softmax.cc` line 93 at r3](https://reviewable.io/reviews/flexflow/flexflow/1106#gh-1362461955:-NgzRUFQCRdIj-2C0Djs:b-896fix) ([raw file](https://github.com/flexflow/flexflow/blob/8c424a967f1ddefc552b134eb31937c348c7195f/lib/runtime/src/ops/softmax.cc#L93)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

```suggestion
  auto input = acc.get_tensor<Permissions::RO>(INPUT);
```
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/softmax.cc` line 122 at r3](https://reviewable.io/reviews/flexflow/flexflow/1106#gh-1362462568:-NgzRc45EORg99wsV_O9:b-896fix) ([raw file](https://github.com/flexflow/flexflow/blob/8c424a967f1ddefc552b134eb31937c348c7195f/lib/runtime/src/ops/softmax.cc#L122)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

```suggestion
  auto output_grad = acc.get_tensor_grad<Permissions::RO>(OUTPUT);
```
</blockquote></details>

Done.

**Review by reyna-abhyankar — APPROVED** [review]

**Pull request merged**